rust/worker/src/segment/types.rs

```diff
@@ -1,29 +1,362 @@
 use std::collections::HashMap;
+use std::sync::atomic::AtomicU32;
+use std::sync::Arc;
 
-use crate::errors::ChromaError;
+use crate::errors::{ChromaError, ErrorCodes};
 use crate::execution::data::data_chunk::Chunk;
-use crate::types::{LogRecord, Metadata};
+use crate::types::{
+    merge_update_metadata, update_metdata_to_metdata, LogRecord, Metadata,
+    MetadataValueConversionError, Operation, OperationRecord,
+};
 use async_trait::async_trait;
+use thiserror::Error;
 
-#[derive(Debug)]
+use super::record_segment::{ApplyMaterializedLogError, RecordSegmentReader};
+
+#[derive(Error, Debug)]
+pub enum LogMaterializerError {
+    #[error("Error materializing document metadata {0}")]
+    MetadataMaterializationError(#[from] MetadataValueConversionError),
+    #[error("Error materializing document embedding")]
+    EmbeddingMaterializationError,
+    #[error("Error reading record segment {0}")]
+    RecordSegmentError(#[from] Box<dyn ChromaError>),
+}
+
+impl ChromaError for LogMaterializerError {
+    fn code(&self) -> ErrorCodes {
+        match self {
+            LogMaterializerError::MetadataMaterializationError(e) => e.code(),
+            LogMaterializerError::EmbeddingMaterializationError => ErrorCodes::Internal,
+            LogMaterializerError::RecordSegmentError(e) => e.code(),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
 pub(crate) struct MaterializedLogRecord<'a> {
```

Comment:
```suggestion
pub(crate) struct MaterializedLogRecord<'referred_data> {
```

prefer explicit named lifetimes